### DISCUSSION TITLE
How to use WiX v4 with CMake

### DISCUSSION THREAD

**Discussion by are-scenic** [wixtoolset/issues#7894] [Questions] [answered]
Currently `cpack` of CMAKE is expecting `candle.exe` installed on the system, which is no longer available in v4. Is there a workaround?
I understood that the different tools of WiX from V3 have been merged into a single tool in V4. Have the arguments of the command-line been preserved?

**Comment by chrpai** [comment, marked as answer]
WiX v3  candle / light  was C++ inspired.  WiX v4 is dotnet(.exe) inspired.  Use the command `dotnet tool install --global wix` to install it and then type wix.  You'll see the commands are different.  `wix build -h` will probably tell you most of what you are looking for.

Also realize that the WiX language (and schema) has changed substantially in v4.  So the CPack WiX generator will need to be updated both in the XML it generates and the commands it uses to build the MSI.   I imagine this to be substantial work for them and they are a year behind if they intend to do it as WiX v4 preview 1 was released a year ago.

**Reply by robmen** [reply]
The changes from v3 to v4 are not that drastic. They definitely are breaking changes but not *that* hard to adapt.

**Reply by chrpai** [reply]
I said substantially not drastic. I was able to update my tooling to handle both the new and old formats but I have no idea how the wix generator in cpack is implemented and what their knowledge level of wix v4 and wix in general is.   For an experienced wix setup developer they aren't that hard to learn the new syntax but for a setup tools developer changing the namespace and root document element structure introduces quite a bit of refactoring.

**Comment by computerquip-work** [comment]
During a move to WiX v4, I separated the actual binary build and the installer build to avoid using CMake and WiX together. CMake outputs an install tarball, WiX just pulls files from that extracted tarball. This requires a bit more effort in the way of hand-writing XML but for simple use-cases, it's not too bad.

**Reply by chrpai** [reply]
I maintain an open source project called IsWiX.  Google IsWiX Tutorials.  I'd be happy to give you an hour of my time to demo it and answer any questions you have.
